File: crates/facet_generate/src/generation/tests/with_enums_across_namespaces/snapshots/root_to_kit/csharp/Example/kit/Kit.cs
using CommunityToolkit.Mvvm.ComponentModel;
using Facet.Runtime.Serde;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using Facet.Runtime.Bincode;

namespace Example.Kit;

public partial class Badge : ObservableObject, IFacetSerializable, IFacetDeserializable<Badge> {
    [ObservableProperty]
    private Presence _presence;
    [ObservableProperty]
    private Shape _shape;

    public void Serialize(ISerializer serializer)
    {
        serializer.IncreaseContainerDepth();
        PresenceBincode.Serialize(Presence, serializer);
        Shape.Serialize(serializer);
        serializer.DecreaseContainerDepth();
    }

    public static Badge Deserialize(IDeserializer deserializer)
    {
        deserializer.IncreaseContainerDepth();
        var presence = PresenceBincode.Deserialize(deserializer);
        var shape = Shape.Deserialize(deserializer);
        deserializer.DecreaseContainerDepth();
        return new Badge {
            Presence = presence,
            Shape = shape,
        };
    }

    public byte[] BincodeSerialize()
    {
        var serializer = new BincodeSerializer();
        Serialize(serializer);
        return serializer.GetBytes();
    }

    public static Badge BincodeDeserialize(byte[] input)
    {
        if (input is null)
        {
            throw new DeserializationError("Cannot deserialize null array");
        }
        var deserializer = new BincodeDeserializer(input);
        var value = Deserialize(deserializer);
        if (deserializer.GetBufferOffset() < input.Length)
        {
            throw new DeserializationError("Some input bytes were not read");
        }
        return value;
    }
}

public enum Presence {
    Online,
    Offline
}

/// <summary>
/// Bincode serialization helpers for <see cref="Presence"/>.
/// </summary>
public static class PresenceBincode {
    public static void Serialize(Presence value, ISerializer serializer)
    {
        serializer.IncreaseContainerDepth();
        serializer.SerializeVariantIndex((uint)value);
        serializer.DecreaseContainerDepth();
    }

    public static Presence Deserialize(IDeserializer deserializer)
    {
        deserializer.IncreaseContainerDepth();
        var index = deserializer.DeserializeVariantIndex();
        deserializer.DecreaseContainerDepth();
        return index switch
        {
            0 => Presence.Online,
            1 => Presence.Offline,
            _ => throw new DeserializationError("Unknown variant index for Presence: " + index),
        }
        ;
    }

    public static byte[] BincodeSerialize(Presence value)
    {
        var serializer = new BincodeSerializer();
        Serialize(value, serializer);
        return serializer.GetBytes();
    }

    public static Presence BincodeDeserialize(byte[] input)
    {
        if (input is null)
        {
            throw new DeserializationError("Cannot deserialize null array");
        }
        var deserializer = new BincodeDeserializer(input);
        var value = Deserialize(deserializer);
        if (deserializer.GetBufferOffset() < input.Length)
        {
            throw new DeserializationError("Some input bytes were not read");
        }
        return value;
    }
}

public abstract record Shape : IFacetSerializable, IFacetDeserializable<Shape> {
    public sealed partial record Circle(double Value) : Shape;

    public sealed partial record Empty() : Shape;

    public abstract void Serialize(ISerializer serializer);

    private static Shape DeserializeCircle(IDeserializer deserializer)
    {
        var value = deserializer.DeserializeF64();
        return new Circle(value);
    }

    public sealed partial record Circle
    {
        public override void Serialize(ISerializer serializer)
        {
            serializer.IncreaseContainerDepth();
            serializer.SerializeVariantIndex(0);
            serializer.SerializeF64(Value);
            serializer.DecreaseContainerDepth();
        }

    }
    private static Shape DeserializeEmpty(IDeserializer deserializer)
    {
        return new Empty();
    }

    public sealed partial record Empty
    {
        public override void Serialize(ISerializer serializer)
        {
            serializer.IncreaseContainerDepth();
            serializer.SerializeVariantIndex(1);
            serializer.DecreaseContainerDepth();
        }

    }
    public static Shape Deserialize(IDeserializer deserializer)
    {
        var index = deserializer.DeserializeVariantIndex();
        return index switch
        {
            0 => DeserializeCircle(deserializer),
            1 => DeserializeEmpty(deserializer),
            _ => throw new DeserializationError("Unknown variant index for Shape: " + index),
        }
        ;
    }

    public byte[] BincodeSerialize()
    {
        var serializer = new BincodeSerializer();
        Serialize(serializer);
        return serializer.GetBytes();
    }

    public static Shape BincodeDeserialize(byte[] input)
    {
        if (input is null)
        {
            throw new DeserializationError("Cannot deserialize null array");
        }
        var deserializer = new BincodeDeserializer(input);
        var value = Deserialize(deserializer);
        if (deserializer.GetBufferOffset() < input.Length)
        {
            throw new DeserializationError("Some input bytes were not read");
        }
        return value;
    }
}

File: crates/facet_generate/src/generation/tests/with_enums_across_namespaces/snapshots/root_to_kit/kotlin/com/example/kit/Kit.kt
package com.example.kit

import com.novi.bincode.BincodeDeserializer
import com.novi.bincode.BincodeSerializer
import com.novi.serde.DeserializationError
import com.novi.serde.Deserializer
import com.novi.serde.Serializer

data class Badge(
    val presence: com.example.kit.Presence,
    val shape: com.example.kit.Shape,
) {
    fun serialize(serializer: Serializer) {
        serializer.increase_container_depth()
        presence.serialize(serializer)
        shape.serialize(serializer)
        serializer.decrease_container_depth()
    }

    fun bincodeSerialize(): ByteArray {
        val serializer = BincodeSerializer()
        serialize(serializer)
        return serializer.get_bytes()
    }

    companion object {
        fun deserialize(deserializer: Deserializer): Badge {
            deserializer.increase_container_depth()
            val presence = com.example.kit.Presence.deserialize(deserializer)
            val shape = com.example.kit.Shape.deserialize(deserializer)
            deserializer.decrease_container_depth()
            return Badge(presence, shape)
        }

        @Throws(DeserializationError::class)
        fun bincodeDeserialize(input: ByteArray?): Badge {
            if (input == null) {
                throw DeserializationError("Cannot deserialize null array")
            }
            val deserializer = BincodeDeserializer(input)
            val value = deserialize(deserializer)
            if (deserializer.get_buffer_offset() < input.size) {
                throw DeserializationError("Some input bytes were not read")
            }
            return value
        }
    }
}

enum class Presence {
    ONLINE,
    OFFLINE;

    fun serialize(serializer: Serializer) {
        serializer.increase_container_depth()
        serializer.serialize_variant_index(ordinal)
        serializer.decrease_container_depth()
    }

    fun bincodeSerialize(): ByteArray {
        val serializer = BincodeSerializer()
        serialize(serializer)
        return serializer.get_bytes()
    }

    companion object {
        @Throws(DeserializationError::class)
        fun deserialize(deserializer: Deserializer): Presence {
            deserializer.increase_container_depth()
            val index = deserializer.deserialize_variant_index()
            deserializer.decrease_container_depth()
            return when (index) {
                0 -> ONLINE
                1 -> OFFLINE
                else -> throw DeserializationError("Unknown variant index for Presence: $index")
            }
        }

        @Throws(DeserializationError::class)
        fun bincodeDeserialize(input: ByteArray?): Presence {
            if (input == null) {
                throw DeserializationError("Cannot deserialize null array")
            }
            val deserializer = BincodeDeserializer(input)
            val value = deserialize(deserializer)
            if (deserializer.get_buffer_offset() < input.size) {
                throw DeserializationError("Some input bytes were not read")
            }
            return value
        }
    }
}

sealed interface Shape {
    fun serialize(serializer: Serializer)

    fun bincodeSerialize(): ByteArray {
        val serializer = BincodeSerializer()
        serialize(serializer)
        return serializer.get_bytes()
    }

    data class Circle(
        val value: Double,
    ) : Shape {
        override fun serialize(serializer: Serializer) {
            serializer.increase_container_depth()
            serializer.serialize_variant_index(0)
            serializer.serialize_f64(value)
            serializer.decrease_container_depth()
        }

        companion object {
            fun deserialize(deserializer: Deserializer): Circle {
                deserializer.increase_container_depth()
                val value = deserializer.deserialize_f64()
                deserializer.decrease_container_depth()
                return Circle(value)
            }
        }
    }

    data object Empty: Shape {
        override fun serialize(serializer: Serializer) {
            serializer.increase_container_depth()
            serializer.serialize_variant_index(1)
            serializer.decrease_container_depth()
        }

        fun deserialize(deserializer: Deserializer): Empty {
            return Empty
        }
    }

    companion object {
        @Throws(DeserializationError::class)
        fun deserialize(deserializer: Deserializer): Shape {
            val index = deserializer.deserialize_variant_index()
            return when (index) {
                0 -> Circle.deserialize(deserializer)
                1 -> Empty.deserialize(deserializer)
                else -> throw DeserializationError("Unknown variant index for Shape: $index")
            }
        }

        @Throws(DeserializationError::class)
        fun bincodeDeserialize(input: ByteArray?): Shape {
            if (input == null) {
                throw DeserializationError("Cannot deserialize null array")
            }
            val deserializer = BincodeDeserializer(input)
            val value = deserialize(deserializer)
            if (deserializer.get_buffer_offset() < input.size) {
                throw DeserializationError("Some input bytes were not read")
            }
            return value
        }
    }
}
